File: src/09.tree/28.112.go
package tree

func HasPathSum(root *TreeNode, targetSum int) bool {
	if root == nil {
		return false
	}

	if root.Left == nil && root.Right == nil {
		return root.Val == targetSum
	}

	return HasPathSum(root.Left, targetSum-root.Val) || HasPathSum(root.Right, targetSum-root.Val)
}

File: src/09.tree/28.112.ts
import { TreeNode } from "./01.100";

export function hasPathSum1(root: TreeNode | null, targetSum: number): boolean {
    let isExist = false;
    const dfs = (node: TreeNode | null, cur: number) => {
        if (!node || isExist) {
            return;
        }

        cur += node.val;

        if (!node.left && !node.right && cur === targetSum) {
            isExist = true;
        }

        dfs(node.left, cur);
        dfs(node.right, cur);
    };

    dfs(root, 0);

    return isExist;
}

export function hasPathSum(root: TreeNode | null, targetSum: number): boolean {
    if (!root) {
        return false;
    }

    if (!root.left && !root.right) {
        return targetSum === root.val;
    }

    return (
        hasPathSum(root.left, targetSum - root.val) ||
        hasPathSum(root.right, targetSum - root.val)
    );
}
